---
title: "DataAnalysis"
author: "Wisse Schuuring"
date: "2/15/2022"
output: pdf_document
---

```{r setup, include=FALSE}

library(ggplot2)
library(dplyr)
library(xtable)


```

# Exploratory Data Analysis


```{r}
#reading in the parental sample files.
sample_1_p <- read.csv(file = "Data/SamplesTSV/GSM3733674_sample1.tsv",
                       sep = "\t", header=T, skip = 3)
sample_4_p <- read.csv(file = "Data/SamplesTSV/GSM3733677_sample4.tsv",
                       sep = "\t", header=T, skip = 3)
sample_5_p <- read.csv(file = "Data/SamplesTSV/GSM3733678_sample5.tsv",
                       sep = "\t", header=T, skip = 3)

#reading in the IRF2 knockout files.
sample_2_KO <- read.csv(file = "Data/SamplesTSV/GSM3733675_sample2.tsv",
                        sep = "\t", header=T, skip = 3)
sample_3_KO <- read.csv(file = "Data/SamplesTSV/GSM3733676_sample3.tsv",
                        sep = "\t", header=T, skip = 3)
sample_6_KO <- read.csv(file = "Data/SamplesTSV/GSM3733679_sample6.tsv",
                        sep = "\t", header=T, skip = 3)


#removing the "values" column in both parental and knockout files.
s1p <- sample_1_p[c("ID_REF", "count")]
s4p <- sample_4_p[c("ID_REF", "count")]
s5p <- sample_5_p[c("ID_REF", "count")]

s2k <- sample_2_KO[c("ID_REF", "count")]
s3k <- sample_3_KO[c("ID_REF", "count")]
s6k <- sample_6_KO[c("ID_REF", "count")]
```

Read in the files and removed the unnecessary value value, for the only ones of interest are the id and the counts value.


```{r}
# Merge the parental data
first_merge_parental <- merge(s1p, s4p, by="ID_REF")
final_merge_parental <- merge(first_merge_parental, s5p, by="ID_REF")
names(final_merge_parental)[2] <- c("counts_1_p")
names(final_merge_parental)[3] <- c("counts_4_p")
names(final_merge_parental)[4] <- c("counts_5_p")
final_merge_parental$counts_total_p <- rowSums(final_merge_parental[2:4])


# Merge the IRF2KO data
first_merge_IRF2KO <- merge(s2k, s3k, by="ID_REF")
final_merge_IRF2KO <- merge(first_merge_parental, s6k, by="ID_REF")
names(final_merge_IRF2KO)[2] <- c("counts_2_KO")
names(final_merge_IRF2KO)[3] <- c("counts_3_KO")
names(final_merge_IRF2KO)[4] <- c("counts_6_KO")
final_merge_IRF2KO$counts_total_KO <- rowSums(final_merge_IRF2KO[2:4])


# Merge both datasets into a completed dataset
MyData <- merge(final_merge_parental, final_merge_IRF2KO, by="ID_REF")

# Write the merged dataset to a seperate file " MyData.csv".
write.csv(MyData,file = 'Data/Samples/Reformat/MyData.csv', row.names=F)

```

With the data finally cleaned up, a simple read.table() shall reveal its contents.

## Reading in the Data

```{r}
# Reading the first five rows of MyData.csv file into R as a table.
read.table(file="Data/Samples/Reformat/MyData.csv", sep=",",
           header=T, nrows = 5)
```

## Data Dimensions

The data set contains the identifier for the gene, followed by the amount of times this gene has been expressed within it's respective sample. In addition, there are "total" columns with the sum of these counts for the IRF2 Knockout samples, and the parental samples. These can be used for comparison's sake. All of these values are integers, therefore easy to work with.

```{r}
# Reading the data set as MyData
MyData <- read.csv(file = "Data/Samples/Reformat/MyData.csv", header = T)

#Show the amount of rows and columns in MyData
Dimensions <- dim(MyData)
cat("The data set contains", Dimensions[1],
    "rows and",Dimensions[2],"columns.\n")

#Show the structure of the data set.
str(MyData)
cat("All values within the data set are integers,
    as expected from a numeric value such as a count.\n")
```

## Creating R Objects

```{r}
# create R objects to store classification
counts_p <- 2:4
counts_KO <- 6:8
counts_total <- c(5,9)
compare_1_2 <- c(2,6)
compare_4_3 <- c(3,7)
compare_5_6 <- c(4,8)
ref_out <- MyData[c(2:4,6:8)]

```

Six R objects are created. The counts objects to keep track of individual counts, as well as a comparison between both totals. The compare objects compare one specific parental expression count value with one specific IRF2 Knockout expression count value. 

## Summarising and Boxplot

```{r}
# Summarise the Data
summary(MyData)

# Create a boxplot displaying the logged data of the six samples.
boxplot(log2(ref_out + 1),
        data=MyData, main="Gene expression between Parental and IRF2 KO",
        xlab="Gene Samples",
        ylab="Log of Gene expressions",
        col = c("orange","orange",
                "orange","dark Blue",
                "dark Blue","dark Blue"),
        las=2, par(cex.axis=0.46),
        names=c("Parent 1","Parent 2",
                "Parent 3","Knockout 1",
                "Knockout 2", "Knockout 3"))
```

A summary shows that most data points do indeed equal zero. The following boxplot reveals however that the data samples overall are very balanced, as they should be, and can all be used for the research. 

## Density Plot

```{r}
# Import affy to create a Density Plot
library(affy)

## Create a vector of 2 colors to use.
library(scales)
myColors <- c("Orange","dark Blue")

## Plot the log2-transformed data with a 0.1 pseudocount
plotDensity(log2(ref_out + 0.1), col=rep(myColors, each=3),
            lty=c(1:ncol(ref_out)), xlab='Log2(count)',
            main='Expression Distribution')

## Add a legend and vertical line
legend('topright', names(ref_out), lty=c(1:ncol(ref_out)),
       col=rep(myColors, each=3))
abline(v=-1.5, lwd=1, col='red', lty=2)
```
Considering there are barely to no shifts between the peaks of the different samples, it can be concluded that the data is of high quality.

## Normalisation

```{r}
## Created a boxplot using MDS
barplot(colSums(ref_out) / 1e6, las = 2,
        col=rep(myColors, each=3),
        xlab = "Gene Samples",
        ylab = "Sequencing Depth (in millions)",
        names=c("Parent 1","Parent 2","Parent 3",
                "Knockout 1","Knockout 2", "Knockout 3"),
        cex.names=0.45)
```

```{r}
library('DESeq2')

# DESeq2 will construct a SummarizedExperiment object and combine this 
# into a 'DESeqDataSet' object. The 'design' argument usually indicates the 
# experimental design using the condition(s) names as a 'factor', for now we use just '~ 1'
(ddsMat <- DESeqDataSetFromMatrix(countData = ref_out,
                                  colData = data.frame(samples = names(ref_out)),
                                  design = ~ 1))
```
```{r}
# Perform normalization
rld.dds <- vst(ddsMat)
# 'Extract' normalized values
rld <- assay(rld.dds)
```

## Distance Calculation

```{r}
# Calculate basic distance metric (using euclidean distance, see '?dist')
sampledists <- dist( t( rld ))
```

## Heatmap

```{r}
# We use the 'pheatmap' library (install with install.packages('pheatmap'))
library(pheatmap)

# Convert the 'dist' object into a matrix for creating a heatmap
(sampleDistMatrix <- as.matrix(sampledists))

# The annotation is an extra layer that will be plotted above the heatmap columns
# indicating the cell type
annotation <- data.frame(Cell = factor(rep(1:2, each = 3),
                                          labels = c("Parental", "IRF2 KO")))

# Set the rownames of the annotation dataframe to the sample names (required)
rownames(annotation) <- names(ref_out)

pheatmap(sampleDistMatrix, show_colnames = FALSE,
         annotation_col = annotation,
         clustering_distance_rows = sampledists,
         clustering_distance_cols = sampledists,
         main = "Euclidean Sample Distances")
```
## Multi-Dimensional Scaling (MDS)

```{r}
library('PoiClaClu')
# Use the raw (not r-log transformed!) counts
dds <- assay(ddsMat)
poisd <- PoissonDistance( t(dds) )
# Extract the matrix with distances
samplePoisDistMatrix <- as.matrix(poisd$dd)
# Calculate the MDS and get the X- and Y-coordinates
mdsPoisData <- data.frame( cmdscale(samplePoisDistMatrix) )

# And set some better readable names for the columns
names(mdsPoisData) <- c('x_coord', 'y_coord')

# Separate the annotation factor (as the variable name is used as label)
groups <- factor(rep(1:2, each=3), 
                 labels = c("Parental", "IRF2 KO"))
coldata <- names(ref_out)

# Create the plot using ggplot
ggplot(mdsPoisData, aes(x_coord, y_coord, color = groups, label = coldata)) + 
  geom_text(size = 4) +
  ggtitle('Multi Dimensional Scaling') +
  labs(x = "Poisson Distance", y = "Poisson Distance") +
  theme_bw()
```
Samples 1 & 2 as well as 3 & 4 are not that different, however there seems to be a large difference between the parental sample 5 and IRF2 KO sample 6.

```{r}
MyData
```
As this dataset only contains 3 samples of each group, as well as the fact the data itself does not contain any missing values, it can be fuly utilised for research. 
