---
title: "DataAnalysis"
author: "Wisse Schuuring"
date: "2/15/2022"
output: pdf_document
---

```{r setup, message=FALSE}
knitr::opts_chunk$set(cache=TRUE)
library(ggplot2)
library(dplyr)
library(xtable)
library(PoiClaClu)
library(affy)
library(scales)
library(pheatmap)
library(DESeq2)
library(pander)

```

# Exploratory Data Analysis


```{r}
#reading in the parental sample files.
sample1P <- read.csv(file = "Data/SamplesTSV/GSM3733674_sample1.tsv",
                       sep = "\t", header=T, skip = 3)
sample4P <- read.csv(file = "Data/SamplesTSV/GSM3733677_sample4.tsv",
                       sep = "\t", header=T, skip = 3)
sample5P <- read.csv(file = "Data/SamplesTSV/GSM3733678_sample5.tsv",
                       sep = "\t", header=T, skip = 3)

#reading in the IRF2 knockout files.
sample2KO <- read.csv(file = "Data/SamplesTSV/GSM3733675_sample2.tsv",
                        sep = "\t", header=T, skip = 3)
sample3KO <- read.csv(file = "Data/SamplesTSV/GSM3733676_sample3.tsv",
                        sep = "\t", header=T, skip = 3)
sample6KO <- read.csv(file = "Data/SamplesTSV/GSM3733679_sample6.tsv",
                        sep = "\t", header=T, skip = 3)


#removing the "values" column in both parental and knockout files.
s1P <- sample1P[c("ID_REF", "count")]
s4P <- sample4P[c("ID_REF", "count")]
s5P <- sample5P[c("ID_REF", "count")]

s2K <- sample2KO[c("ID_REF", "count")]
s3K <- sample3KO[c("ID_REF", "count")]
s6K <- sample6KO[c("ID_REF", "count")]
```

Read in the files and removed the unnecessary value value, for the only ones of interest are the id and the counts value.


```{r}
# Merge the parental data
firstMergeP <- merge(s1P, s4P, by="ID_REF")
finalMergeP <- merge(firstMergeP, s5P, by="ID_REF")
names(finalMergeP)[2] <- c("counts_1_p")
names(finalMergeP)[3] <- c("counts_4_p")
names(finalMergeP)[4] <- c("counts_5_p")


# Merge the IRF2KO data
firstMergeIRF2KO <- merge(s2K, s3K, by="ID_REF")
finalMergeIRF2KO <- merge(firstMergeIRF2KO, s6K, by="ID_REF")
names(finalMergeIRF2KO)[2] <- c("counts_2_KO")
names(finalMergeIRF2KO)[3] <- c("counts_3_KO")
names(finalMergeIRF2KO)[4] <- c("counts_6_KO")


# Merge both datasets into a completed dataset
myData <- merge(finalMergeP, finalMergeIRF2KO, by="ID_REF")

# Write the merged dataset to a seperate file " myData.csv".
write.csv(myData,file = 'Data/Samples/Reformat/myData.csv', row.names=F)

```

With the data finally cleaned up, a simple read.table() shall reveal its contents.

## Reading in the Data

```{r}
# Reading the first five rows of myData.csv file into R as a table.
read.table(file="Data/Samples/Reformat/myData.csv", sep=",",
           header=T, row.names=1, nrows = 5)
```

## Data Dimensions

The data set contains the identifier for the gene, followed by the amount of times this gene has been expressed within it's respective sample. In addition, there are "total" columns with the sum of these counts for the IRF2 Knockout samples, and the parental samples. These can be used for comparison's sake. All of these values are integers, therefore easy to work with.

```{r}
# Reading the data set as myData
myData <- read.csv(file = "Data/Samples/Reformat/myData.csv", header = T, row.names = 1)

#Show the amount of rows and columns in myData
dimensions <- dim(myData)
cat("The data set contains", dimensions[1],
    "rows and",dimensions[2],"columns.\n")

#Show the structure of the data set.
pander(str(myData))
cat("All values within the data set are integers,
    as expected from a numeric value such as a count.\n")
```

## Creating R Objects

```{r}
# create R objects to store classification
countsP <- 1:3
countsKO <- 4:6
compare1To2 <- c(1,4)
compare4To3 <- c(2,5)
compare5To6 <- c(3,5)

```

Six R objects are created. The counts objects to keep track of individual counts, as well as a comparison between both totals. The compare objects compare one specific parental expression count value with one specific IRF2 Knockout expression count value. 

## Summarising and Boxplot

```{r}
# Summarise the Data
pander(summary(myData))

# Create a boxplot displaying the logged data of the six samples.
boxplot(log2(myData + 1),
        data=myData, main="Gene expression between Parental and IRF2 KO",
        xlab="Gene Samples",
        ylab="Log of Gene expressions",
        col = c("orange","orange",
                "orange","dark Blue",
                "dark Blue","dark Blue"),
        las=2, par(cex.axis=0.46),
        names=c("Parent 1","Parent 2",
                "Parent 3","Knockout 1",
                "Knockout 2", "Knockout 3"))
```

A summary shows that most data points do indeed equal zero. The following boxplot reveals however that the data samples overall are very balanced, as they should be, and can all be used for the research. 

## Density Plot

```{r}
## Create a vector of 2 colors to use.
myColors <- c("Orange","dark Blue")

## Plot the log2-transformed data with a 0.1 pseudocount
plotDensity(log2(myData + 0.1), col=rep(myColors, each=3),
            lty=c(1:ncol(myData)), xlab='Log2(count)',
            main='Expression Distribution')

## Add a legend and vertical line
legend('topright', names(myData), lty=c(1:ncol(myData)),
       col=rep(myColors, each=3))
abline(v=-1.5, lwd=1, col='red', lty=2)
```
Considering there are barely to no shifts between the peaks of the different samples, it can be concluded that the data is of high quality.

## Normalisation

```{r}
## Created a boxplot using MDS
barplot(colSums(myData) / 1e6, las = 2,
        col=rep(myColors, each=3),
        main = "Sequencing Depth",
        xlab = "Gene Samples",
        ylab = "Sequencing Depth (in millions)",
        names=c("Parent 1","Parent 2","Parent 3",
                "Knockout 1","Knockout 2", "Knockout 3"),
        cex.names=0.45)
```

```{r}

# DESeq2 will construct a SummarizedExperiment object and combine this 
# into a 'DESeqDataSet' object. The 'design' argument usually indicates the 
# experimental design using the condition(s) names as a 'factor',
# for now we use just '~ 1'
(ddsMat <- DESeqDataSetFromMatrix(countData = myData,
                                  colData = data.frame(samples = names(myData)),
                                  design = ~ 1))
```
```{r}
# Perform normalization
rld.dds <- vst(ddsMat)
# 'Extract' normalized values
rld <- assay(rld.dds)
```

## Distance Calculation

```{r}
# Calculate basic distance metric (using euclidean distance, see '?dist')
sampledists <- dist( t( rld ))
```

## Heatmap

```{r}

# Convert the 'dist' object into a matrix for creating a heatmap
(sampleDistMatrix <- as.matrix(sampledists))

# The annotation is an extra layer that will be plotted above the heatmap columns
# indicating the cell type
annotation <- data.frame(Type = factor(rep(1:2, each = 3),
                                          labels = c("Parental", "IRF2 KO")))

# Set the rownames of the annotation dataframe to the sample names (required)
rownames(annotation) <- names(myData)

pheatmap(sampleDistMatrix, show_colnames = FALSE,
         annotation_col = annotation,
         clustering_distance_rows = sampledists,
         clustering_distance_cols = sampledists,
         main = "IRF2 Knockout Sample Distances")
```
## Multi Dimensional Scaling (MDS)

```{r}
# Use the raw (not r-log transformed!) counts
dds <- assay(ddsMat)
poisd <- PoissonDistance( t(dds) )
# Extract the matrix with distances
samplePoisDistMatrix <- as.matrix(poisd$dd)
# Calculate the MDS and get the X- and Y-coordinates
mdsPoisData <- data.frame( cmdscale(samplePoisDistMatrix) )

# And set some better readable names for the columns
names(mdsPoisData) <- c('x_coord', 'y_coord')

# Separate the annotation factor (as the variable name is used as label)
groups <- factor(rep(1:2, each=3), 
                 labels = c("Parental", "IRF2 KO"))
coldata <- names(myData)

# Create the plot using ggplot
ggplot(mdsPoisData, aes(x_coord, y_coord, color = groups, label = coldata)) + 
  geom_text(size = 4) +
  ggtitle('Multi Dimensional Scaling') +
  labs(x = "Poisson Distance", y = "Poisson Distance") +
  theme_bw()
```
There is a clear poisson distance between the parental and the IRF2 KO gene samples. As this data set only contains 3 samples of each group, as well as the fact the data itself does not contain any missing values, it can be fully utilized for research. 

# Discovering Differentialy Expressed Genes

FILTER OUT LOW COUNT DATA



```{r}

myData <- read.table(file="Data/Samples/Reformat/myData.csv", sep=",",
           header=T, row.names=1)

# Perform a naive FPM normalization
# Note: log transformation includes a pseudocount of 1

counts.fpm <- log2( (myData / (colSums(myData) / 1e6)) + 1 )

#if ever value of a row have values of below five, remove them
row_sub <- apply(counts.fpm, 1, function(row) all(row > 5 ))
row_removed <- apply(counts.fpm, 1, function(row) any(row <= 5))

# Show the identifiers of the genes that were removed due to containing 0 values
removed_genes <- counts.fpm[row_removed,]

# Show the remaining genes and their count values
genes <- counts.fpm[row_sub,]

totalAmount <- dplyr::count(myData)[[1]]
keepAmount <- dplyr::count(genes)[[1]]
removeAmount <- dplyr::count(removed_genes)[[1]]

keepPercent <- round((keepAmount*100)/totalAmount)
removedPercent <- round((removeAmount*100)/totalAmount)

cat("The total amount of genes in the data is",totalAmount,
    ".\n Of those genes,",keepAmount,
    " were above a count of zero,\n with",removeAmount,
    " equal to it or below. In conclusion,\n",keepPercent,
    "% of the data will be kept, while",removedPercent,"% will be removed.")
  
```

CALCULATE THE FOLD CHANGE VALUE

# The Fold Change Value (FC)


```{r}
# Create a function that performs a t.test per row in a data frame.
myTTest <- function(y){
  return(t.test(y[1:3], y[4:6])$p.value)
}

# assign the row and column names
gene.names <- rownames(genes)
samples <- colnames(genes)

# Apply the function to the logged data set.
pVec <- apply(genes, 1, myTTest)
alpha <- 0.05

#Show how many DEGs have been found.
cat("Number of DEGs:\n")
( n.DEG <- sum(pVec < alpha) )
```



```{r}
# Create two columns, one with means of parental and one with Knockout.
genes$avg_p <- rowMeans(genes[1:3])
genes$avg_KO <- rowMeans(genes[4:6])

# Calculate the FC values.
genes$foldChange <- (genes$avg_KO - genes$avg_p)

# Display the values as a histogram.
hist(as.matrix(genes$foldChange), breaks=80, main="Fold Change Value of the Gene Counts", xlab="FC value", col= rgb(0.5,0,0.8,1/2))
abline(v = c(-1,1), col="red")
```
# Using Bioconductor Packages

"Differential expression analysis was performed using the voom/limma R package
(38). Genes were considered to be differentially expressed if the log2
of the fold change was >1 or < 1 and the adjusted P value was <0.05"


## Limma package

```{r}
library(edgeR)

# Read in the file containing the count data.
counts <- read.csv("Data/Samples/Reformat/myData.csv", row.names = 1)
head(counts)

# Create DEGList object
d0 <- DGEList(counts, group = c(1,1,1,2,2,2))

#Calculate normalization factors
d0 <- calcNormFactors(d0)
```

```{r}

# Filter low-expressed genes
cutoff <- 5
drop <- which(apply(cpm(d0), 1, max) < cutoff)
d <- d0[-drop,] 

dim(d) # number of genes left
```

```{r}
# Multidimensional scaling (MDS) plot
group <- factor(c(1,1,1,2,2,2), labels = c("Control", "Case"))
plotMDS(d, col = as.numeric(group))
```
```{r}
# Voom transformation and calculation of variance weights.
# Specify the model to be fitted. We do this before using voom since voom uses variances of the model residuals (observed - fitted)
mm <- model.matrix(~0 + group)

y <- voom(d, mm, plot = T)
```
1. Counts are transformed to log2 counts per million reads (CPM),
where “per million reads” is defined based on the normalization
factors we calculated earlier

2. A linear model is fitted to the log2 CPM for each gene,
and the residuals are calculated

3. A smoothed curve is fitted to the
sqrt(residual standard deviation) by average expression
(see red line in plot above)

4. The smoothed curve is used to obtain weights for each gene
and sample that are passed into limma along with the log2 CPMs.

```{r}
# Fitting linear models in limma

fit <- lmFit(y, mm)
head(coef(fit))

# Comparison between control (parental) and case (KO)
contr <- makeContrasts(groupCase - groupControl, levels = colnames(coef(fit)))
contr

# Estimate contrast for each gene
tmp <- contrasts.fit(fit, contr)

# Empirical Bayes smoothing of standard errors 
tmp <- eBayes(tmp)

# What genes are most differentially expressed?
top.table <- topTable(tmp, sort.by = "P", n = Inf)
head(top.table, 20)
```
logFC: log2 fold change of Control/Case
AveExpr: Average expression across all samples, in log2 CPM
t: logFC divided by its standard error
P.Value: Raw p-value (based on t) from test that logFC differs from 0
adj.P.Val: Benjamini-Hochberg false discovery rate adjusted p-value
B: log-odds that gene is DE (arguably less useful than the other columns)

```{r}
cat("Number of DEGs:\n")
cat(length(which(top.table$logFC < -1 | top.table$logFC > 1)))
```


```{r}
# Display the values as a histogram.
hist(as.matrix(top.table$logFC), breaks=80, main="Fold Change Value of the Gene Counts", xlab="FC value", col= rgb(0.5,0,0.8,1/2))
abline(v = c(-1,1), col="red")
```
## Data Analysis and Visualisation

```{r}
library(EnhancedVolcano)



## Simple function for plotting a Volcano plot, returns a ggplot object
deseq.volcano <- function(res, datasetName) {
  return(EnhancedVolcano(res, x = 'logFC', y = 'adj.P.Val',
                         lab=rownames(res),
                         title = paste(datasetName, "Parental vs Knockout"),
                         subtitle = bquote(italic('adj.P.Val <= 0.05 and absolute FC >= 2')),
                         # Change text and icon sizes
                         labSize = 3, pointSize = 1.5, axisLabSize=10, titleLabSize=12,
                         subtitleLabSize=8, captionLabSize=10,
                         # Disable legend
                         legendPosition = "none",
                         # Set cutoffs
                         pCutoff = 0.05, FCcutoff = 2))
}

## Note: input data is the corrected DESeq2 output using the 'lfcShrink' function (see chapter 4)
deseq.volcano(res = top.table, datasetName = "top")
```
```{r}

y <- DGEList(counts, group = c(1,1,1,2,2,2))

y <- calcNormFactors(y)

y <- estimateDisp(y)

y <- estimateCommonDisp(y)

y <- estimateTagwiseDisp(y)

et <- exactTest(y)
tt <- topTags(et, n = Inf)

edgeR.gene.ids <- row.names(tt$table)[which(abs(tt$table$logFC) > 1 & tt$table$FDR < 0.05)]


```





```{r}
## Venn-diagram using the `VennDiagram` library (see below for alternative method)
library(VennDiagram)

voom.gene.ids <- row.names(top.table)[which(abs(top.table$logFC) > 1 & top.table$adj.P.Val < 0.05)]


# Arguments for a pairwise (two-sets) venn-diagram are sizes for set1, set2 and overlap (intersect)
# Many more functions are available for triple, quad and quantuple diagrams (starting with 'draw.***')
venn.plot <- draw.pairwise.venn(length(voom.gene.ids),
                                length(edgeR.gene.ids),
                                # Calculate the intersection of the two sets
                                length( intersect(voom.gene.ids, edgeR.gene.ids) ),
                                category = c("Voom", "EdgeR"), scaled = T,
                                fill = c("light blue", "pink"), alpha = rep(0.5, 2),
                                cat.pos = c(0, 0))

# Actually plot the plot
grid.draw(venn.plot)
```

