---
title: "Logbook"
output: html_notebook
---

week 1:

searched the website

found old data set

looked at old data set and associated article

determined core research


week 2:

Explored files in data set

created Project Proposal

Began EDA and sorted files into single large file

Upon receiving the data, the samples were split up into six seperated tsv files. first, I had to reformat them into csv. Next, I had to add a column to every csv file, to show which sample it belonged to. Next, I had to place all sample data in a single file. 

```{r}
# rawdata <- read.csv(file = 'Data/Samples/GSM3733674_sample1.csv', sep = ';', header=T)
# rawdata$SAMPLE <- 1
# write.csv(rawdata,file = 'Data/Samples/Reformat/sample1.csv', row.names=F, col.names=F)
# 
# rawdata2 <- read.csv(file = 'Data/Samples/GSM3733675_sample2.csv', sep = ';', header=T)
# rawdata2$SAMPLE <- 2
# write.csv(rawdata2,file = 'Data/Samples/Reformat/sample2.csv', row.names=F, col.names=F)
# 
# rawdata3 <- read.csv(file = 'Data/Samples/GSM3733676_sample3.csv', sep = ';', header=T)
# rawdata3$SAMPLE <- 3
# write.csv(rawdata3,file = 'Data/Samples/Reformat/sample3.csv', row.names=F, col.names=F)
# 
# rawdata4 <- read.csv(file = 'Data/Samples/GSM3733677_sample4.csv', sep = ';', header=T)
# rawdata4$SAMPLE <- 4
# write.csv(rawdata4,file = 'Data/Samples/Reformat/sample4.csv', row.names=F, col.names=F)
# 
# rawdata5 <- read.csv(file = 'Data/Samples/GSM3733678_sample5.csv', sep = ';', header=T)
# rawdata5$SAMPLE <- 5
# write.csv(rawdata5,file = 'Data/Samples/Reformat/sample5.csv', row.names=F, col.names=F)
# 
# rawdata6 <- read.csv(file = 'Data/Samples/GSM3733679_sample6.csv', sep = ';', header=T)
# rawdata6$SAMPLE <- 6
# write.csv(rawdata6,file = 'Data/Samples/Reformat/sample6.csv', row.names=F, col.names=F)
# #
# 
# rawdata <- read.csv(file = 'Data/Samples/Reformat/sample1.csv')
# rawdata$TYPE <- "Parental"
# write.csv(rawdata,file = 'Data/Samples/Reformat/sample_1.csv', row.names=F)
# 
# rawdata2 <- read.csv(file = 'Data/Samples/Reformat/sample2.csv')
# rawdata2$TYPE <- "IRF2_KO"
# write.csv(rawdata2,file = 'Data/Samples/Reformat/sample_2.csv', row.names=F)
# 
# rawdata3 <- read.csv(file = 'Data/Samples/Reformat/sample3.csv')
# rawdata3$TYPE <- "IRF2_KO"
# write.csv(rawdata3,file = 'Data/Samples/Reformat/sample_3.csv', row.names=F)
# 
# rawdata4 <- read.csv(file = 'Data/Samples/Reformat/sample4.csv')
# rawdata4$TYPE <- "Parental"
# write.csv(rawdata4,file = 'Data/Samples/Reformat/sample_4.csv', row.names=F)
# 
# rawdata5 <- read.csv(file = 'Data/Samples/Reformat/sample5.csv')
# rawdata5$TYPE <- "Parental"
# write.csv(rawdata5,file = 'Data/Samples/Reformat/sample_5.csv', row.names=F)
# 
# rawdata6 <- read.csv(file = 'Data/Samples/Reformat/sample6.csv')
# rawdata6$TYPE <- "IRF2_KO"
# write.csv(rawdata6,file = 'Data/Samples/Reformat/sample_6.csv', row.names=F)

# AllData <- read.csv(file="Data/Samples/Reformat/AllSamples.csv", header =T)
# MyData <- AllData[c("ID_REF","count","SAMPLE")]
# MyData

```

Here I attempted to add whether they were parental or IRF2KO, however afterwards I decided placing each in the same file, with the type noted in the column name would be more efficient.

```{r}

# first_merge <- merge(s1p, s2k, by="ID_REF")
# names(first_merge)[2] <- "counts_1_p"
# names(first_merge)[3] <- "counts_2_KO"
# 
# second_merge <- merge(first_merge, s4p, by="ID_REF")
# names(second_merge)[4] <- "counts_4_p"
# 
# third_merge <- merge(second_merge, s3k, by="ID_REF")
# names(third_merge)[5] <- "counts_3_KO"
# 
# fourth_merge <- merge(third_merge, s5p, by="ID_REF")
# names(fourth_merge)[6] <- "counts_5_p"
# 
# final_merge <- merge(fourth_merge, s6k, by="ID_REF")
# names(final_merge)[7] <- "counts_6_KO"

#final_merge

#write.csv(final_merge,file = 'Data/Samples/Reformat/MyData.csv', row.names=F)

```

This merges all sample files together one by one, however for comparisons sake I shall first create two separate files with a sum of the counts for the parental samples and IRF2KO samples respectively. This should make comparing the counts later easier. 

Created Github and pushed the files upon it

Data structure checked for data type using dim() and str()

created R objects from the data set for use in classification


week 3:

Cleaned up the reading in section so I can now read the raw tsv files instead of edited csv files.

Fixed the R objects from their previous wrong assignment

Created a boxplot using previously mentioned R objects

Created a density plot using previously mentioned R objects

Created a barplot displaying MultiDimensional Scaling

Spent too much time adding legends to my plots.

Performed Normalisation of my data

Created a Heatmap

Performed Multi-Dimensional Scaling

fixed the dataset, made a mistake in my merging. 

moved libraries to the correct R chunk and removed unrecognised Unicode characters for pdf Knit.

```{r}
# boxplot(counts_p)
# boxplot(counts_KO)
# count_compare <- ggplot(counts_total, aes(x=counts_total_p, y=counts_total_KO)) + geom_point() +  scale_x_continuous() +
#   geom_jitter()
# count_compare
#ggplot(counts_total, aes(x=counts_total_KO, y=ID_REF))
#boxplot(MyData$counts_total_KO ~ MyData$ID_REF)

summary(MyData)
# boxplot(ref_out, outline = FALSE)
# boxplot(log2(ref_out + 1), outline = FALSE)

boxplot(log2(ref_out + 1),
        data=MyData, main="Gene expression between Parental and IRF2 KO",
        xlab="Gene Samples",
        ylab="Log of Gene expressions",
        col = c("dark green","dark green",
                "dark green","light green",
                "light green","light green"),
        las=3, par(cex.axis=0.46),
        names=c("Parent 1","Parent 2",
                "Parent 3","Knockout 1",
                "Knockout 2", "Knockout 3"))

```

Week 4

Began pre processing by removing all 0 values

created a plot using only the genes used.

week 5

```{r}
# gene_avg <- MyData[c(4,8)]
# avg.fpm <- log2( (gene_avg / (colSums(MyData) / 1e6)) + 1 )
# avg.fpm
# p1 <- hist(avg.fpm$counts_avg_p, col="Orange")
# p2 <- hist(avg.fpm$counts_avg_KO, col="Blue")
# plot( p1, col=rgb(1,0.5,0,1/1.8), main = "Fold Change Values Frequencies")
# plot( p2, col=rgb(0,0,1,1/1.8), add=T)
# legend(8, 2500, legend = c("Parental","Knockout"), fill = c("Orange","Blue"))
```

Succesfully calculated DEG's. 

Created a dataset to hold the DEGs.

Calculated the average value of the DEG counts, logged them, and created a plot to show their differences.

Created a histogram that shows the FC Value of the log transformed data.

Began using the limma package for Differential Expression

Week 6:

Attempted to use DESeq2 for fc calculations

Attempted to use EdgeR for calculations.

settled on Limma only, applied article limitations. 

week 7: 

Created volcano plot displaying DEG's in limma top table data.

week 8: 

Created Venn diagram after many iterations.
